Platform/Authentication/Token Provider/DefaultTokenProvider.swift

```diff
@@ -0,0 +1,228 @@
+    /// to application/x-www-form-urlencoded.
+    public var headersInjector: HeadersInjector?
+    
+    /// An optional closure which will be invoked when a request is about to be made, so that you can
+    /// supply query items which your backend might require to process the request.
+    public var queryItemsInjector: QueryItemsInjector?
+    
+    /// An optional closure which will be invoked when a request is about to be made, so that you can
+    /// supply URL encoded body items which your backend might require to process the request.
+    ///
+    /// If not specified otherwise, the token provider will always add grant_type=client_credentials item
+    /// to the body of the request.
+    public var bodyInjector: BodyInjector?
+    
+    /// An optional logger used by the token provider.
+    public let logger: PPLogger?
+    
+    // MARK: - Initializers
+    
+    /// Create an DefaultTokenProvider which presents either headers, query items or URL encoded body
+    /// items as part of the request. These should be used to identify your application user session to your
+    /// backend so that it can issue a token for the user.
+    ///
+    /// - Parameters:
+    ///     - url: The URL to be called.
+    ///     - headers: An optional dictionary of headers to include in the request. Here you can
+    ///     supply the session or other credientials which your endpoint might require to authenticate
+    ///     the request. If not specified otherwise, the token provider will always set Content-Type header
+    ///     to application/x-www-form-urlencoded.
+    ///     - queryItems: An optional list of query items to include in the request URL. Here you can
+    ///     supply any query items which your endpoint might require to process the request.
+    ///     - body: An optional list of URL encoded body items to include in the request. Here you can
+    ///     supply any body items which your endpoint might require to process the request. If not
+    ///     specified otherwise, the token provider will always add grant_type=client_credentials item
+    ///     to the body of the request.
+    ///     - logger: An optional logger used by the token provider.
+    public init(url: URL, headers: [String : String]? = nil, queryItems: [URLQueryItem]? = nil, body: [URLEncodedBodyItem]? = nil, logger: PPLogger? = nil) {
+        self.url = url
+        self.headers = headers
+        self.queryItems = queryItems
+        self.body = body
+        self.headersInjector = nil
+        self.queryItemsInjector = nil
+        self.bodyInjector = nil
+        self.logger = logger
+    }
+    
+    // MARK: - Token retrieval
+    
+    /// Method called by the SDK to authenticate the user.
+    ///
+    /// - Parameters:
+    ///     - completionHandler: The completion handler that provides
+    ///     `AuthenticationResult` to the SDK.
+    public func fetchToken(completionHandler: @escaping (AuthenticationResult) -> Void) {
```

Comment:
There are currently 12 tests of this method in Functional Tests target, which stubs `URLSession` as you suggested.

Reply:
Your point is totally valid.  Thought about this for a bit and started that convo on Slack as a result.

But I've had a further realisation... I think the 12 tests you're referring to can actually be classified as *unit* tests (rather than functional).  They tests a single entity with all dependencies mocked.  It just so happens that the dependency is the network but I don't think that in iteslf makes the test functional.  

Functional tests are tests that test the system as a whole and thats not really whats happened there.

Reply:
I merged functional tests with unit tests.

Reply:
👍  nice one